File: app/src/main/java/com/laushkina/anastasia/psychosophy/view/test/TestFragment.java
package com.laushkina.anastasia.psychosophy.view.test;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import com.laushkina.anastasia.psychosophy.view.BaseFragment;
import com.laushkina.anastasia.psychosophy.R;
import com.laushkina.anastasia.psychosophy.databinding.FragmentTestBinding;
import com.laushkina.anastasia.psychosophy.domain.Psychotype;
import com.laushkina.anastasia.psychosophy.view.NavigationHelper;

import java.util.List;

import javax.inject.Inject;

public class TestFragment extends BaseFragment implements ITestResultsObserver, ITypeCalculator {

    private static final String TEST_STATE_EXTRA = "test_state";
    @Inject TestPresenter presenter;
    private TestViewModel viewModel;
    private TestQuestionsAdapter adapter;

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState){
        super.onCreateView(inflater, container, savedInstanceState);

        FragmentTestBinding binding = DataBindingUtil.inflate(inflater, R.layout.fragment_test,
                container, false);
        binding.setTestCalculator(this);

        presenter = DaggerTestComponent.create().getPresenter();
        View view = binding.getRoot();
        initialize(view);
        binding.setViewModel(viewModel);

        return view;
    }

    @Override
    public void onSaveInstanceState(Bundle bundle){
        super.onSaveInstanceState(bundle);
        if (isVisible()) {
            bundle.putBundle(TEST_STATE_EXTRA, presenter.getTestState());
        }
    }

    @Override
    public void onViewStateRestored(Bundle bundle) {
        super.onViewStateRestored(bundle);

        Bundle savedState = bundle == null ? null : bundle.getBundle(TEST_STATE_EXTRA);
        if (savedState == null) return;

        presenter.restoreFromSavedState(savedState);
        viewModel.setProgress(presenter.getProgress());
    }

    private void initialize(View view){
        initialize();
        viewModel = new TestViewModel(presenter.getNextQuestionText(this));

        // Initialize questions
        RecyclerView questionsRecycler = getQuestionsRecycler(view);
        questionsRecycler.setLayoutManager(new LinearLayoutManager(getActivity()));
        adapter = new TestQuestionsAdapter(presenter.getFirstGroupOfQuestions(this), this::onAnswersSelected);
        questionsRecycler.setAdapter(adapter);

        getProgressBar(view).setProgressDrawable(getResources().getDrawable(R.drawable.test_progress_drawable));
    }

    void onAnswersSelected() {
        presenter.onAnswerSelected(this);
    }

    public void next() {
        presenter.onNextRequested(adapter.getAnswers(),this);
    }

    public void prev() {
        presenter.onPrevQuestionsRequested(this);
    }

    @Override
    public void showGroupOfQuestions(List<CharSequence> questionSet, boolean isPrevEnabled) {
        adapter.setQuestions(questionSet);
        viewModel.setProgress(presenter.getProgress());
        viewModel.setNextButtonText(presenter.getNextQuestionText(this));
        viewModel.setPrevButtonText(getResources().getString(R.string.prev_question_title));
        viewModel.setPrevEnabled(isPrevEnabled);
        viewModel.setNextEnabled(false);
    }

    @Override
    public void showNextButton() {
        viewModel.setNextEnabled(true);
    }

    @Override
    public void showExceptionResultDescription(){
        NavigationHelper.getInstance().showTestResults(null, getFragmentManager());
    }

    @Override
    public void showTypeDescription(Psychotype[] psychotypes) {
        NavigationHelper.getInstance().showTestResults(psychotypes, getFragmentManager());
    }

    @Override
    public Context getContext() {
        return getActivity();
    }

    @Override
    protected String getTitle(){
        return getResources().getString(R.string.test);
    }

    @Override
    protected int getNavigationItemId(){
        return R.id.nav_test;
    }

    private ProgressBar getProgressBar(View view){
        return view.findViewById(R.id.test_progress);
    }

    private RecyclerView getQuestionsRecycler(View view){
        return view.findViewById(R.id.test_questions_recycler);
    }
}

File: app/src/main/java/com/laushkina/anastasia/psychosophy/view/test/TestFragment.kt
package com.laushkina.anastasia.psychosophy.view.test

import androidx.databinding.DataBindingUtil
import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import com.laushkina.anastasia.psychosophy.R
import com.laushkina.anastasia.psychosophy.databinding.FragmentTestBinding
import com.laushkina.anastasia.psychosophy.domain.Psychotype
import com.laushkina.anastasia.psychosophy.view.BaseFragment
import com.laushkina.anastasia.psychosophy.view.NavigationHelper
import javax.inject.Inject

class TestFragment: BaseFragment(), ITestResultsObserver, ITypeCalculator {

    companion object {
        private const val TEST_STATE_EXTRA = "test_state"
    }

    @Inject
    internal lateinit var presenter: TestPresenter
    private lateinit var viewModel: TestViewModel
    private lateinit var adapter: TestQuestionsAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)

        val binding = DataBindingUtil.inflate<FragmentTestBinding>(inflater, R.layout.fragment_test,
                container, false)
        binding.testCalculator = this

        presenter = DaggerTestComponent.create().getPresenter()
        val view = binding.root
        initialize(view)
        binding.viewModel = viewModel

        return view
    }
    
    override fun onSaveInstanceState(bundle: Bundle) {
        super.onSaveInstanceState(bundle)
        if (isVisible) {
            bundle.putBundle(TEST_STATE_EXTRA, presenter.getTestState())
        }
    }

    override fun onViewStateRestored(bundle: Bundle?) {
        super.onViewStateRestored(bundle)
        val savedState = bundle?.getBundle(TEST_STATE_EXTRA) ?: return

        presenter.restoreFromSavedState(savedState, this, context!!)
    }


    override fun getTitle(): CharSequence {
        return resources.getString(R.string.test)
    }

    override fun getNavigationItemId(): Int {
        return R.id.nav_test
    }

    override fun showExceptionResultDescription() {
        NavigationHelper.instance.showTestResults(null, fragmentManager)
    }

    override fun showTypeDescription(psychotypes: Array<Psychotype>) {
        NavigationHelper.instance.showTestResults(psychotypes, fragmentManager)
    }

    override fun showGroupOfQuestions(questionSet: List<CharSequence>, isPrevEnabled: Boolean, progress: Int, nextButtonText: String) {
        getQuestionsRecycler().scrollToPosition(0)
        adapter.setQuestions(questionSet)
        viewModel.setProgress(progress)
        viewModel.setNextButtonText(nextButtonText)
        viewModel.setPrevButtonText(resources.getString(R.string.prev_question_title))
        viewModel.setPrevEnabled(isPrevEnabled)
        viewModel.setNextEnabled(false)
    }

    override fun showNextButton() {
        viewModel.setNextEnabled(true)
    }

    override fun next() {
        val context = context
        if (context != null) {
            presenter.onNextRequested(adapter.getAnswers(), this, context)
        }
    }

    override fun prev() {
        val context = context
        if (context != null) {
            presenter.onPrevQuestionsRequested(this, context)
        }
    }

    private fun initialize(view: View) {
        initialize()
        viewModel = TestViewModel()
        viewModel.setNextButtonText(presenter.getNextQuestionText(view.context))

        // Initialize questions
        val questionsRecycler = getQuestionsRecycler(view)
        questionsRecycler.layoutManager = LinearLayoutManager(activity)
        adapter = TestQuestionsAdapter(
                presenter.getFirstGroupOfQuestions(context!!),
                object: TestQuestionsAdapter.OnAnswersSelectedListener {
                    override fun onAllAnswersSelected() {
                        presenter.onAnswerSelected(this@TestFragment)
                    }
                }
        )
        questionsRecycler.adapter = adapter

        getProgressBar(view).progressDrawable = resources.getDrawable(R.drawable.test_progress_drawable)
    }

    private fun getProgressBar(view: View): ProgressBar {
        return view.findViewById(R.id.test_progress)
    }

    private fun getQuestionsRecycler(view: View): RecyclerView {
        return view.findViewById(R.id.test_questions_recycler)
    }

    // Must be called after onAttach() and before on detach() or NPE will happen
    private fun getQuestionsRecycler(): RecyclerView {
        return activity!!.findViewById(R.id.test_questions_recycler)
    }

}